plugins/vli/fu-vli-usbhub-realtek-device.c

```diff
@@ -0,0 +1,472 @@
+	guint8 status = 0xff;
+	if (!fu_vli_usbhub_device_realtek_read_status_raw (self, &status, error))
+		return FALSE;
+	if (status == 0xBB) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "status was 0x%02x", status);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_vli_usbhub_device_realtek_read_status (FuVliUsbhubRealtekDevice *self,
+					  guint8 *status,
+					  GError **error)
+{
+	return fu_device_retry (FU_DEVICE (self),
+				fu_vli_usbhub_device_realtek_read_status_cb,
+				4200, status, error);
+}
+
+static gboolean
+fu_vli_usbhub_realtek_ensure_version_unlocked (FuVliUsbhubRealtekDevice *self,
+					      GError **error)
+{
+	FuVliUsbhubDevice *parent = FU_VLI_USBHUB_DEVICE (fu_device_get_parent (FU_DEVICE (self)));
+	guint8 buf_rep[7] = { 0x0 };
+	guint8 buf_req[1] = { 0x04 };
+	g_autofree gchar *version = NULL;
+
+	if (!fu_vli_usbhub_device_i2c_write (parent,
+					     UC_FOREGROUND_SLAVE_ADDR,
+					     UC_FOREGROUND_OPCODE,
+					     buf_req, sizeof(buf_req),
+					     error)) {
+		g_prefix_error (error, "failed to get version number: ");
+		return FALSE;
+	}
+
+	/* FIXME: whoooa */
+	g_usleep (300000);
+	if (!fu_vli_usbhub_device_i2c_read (parent,
+					    UC_FOREGROUND_SLAVE_ADDR,
+					    0x00,
+					    buf_rep, sizeof(buf_rep),
+					    error)) {
+		g_prefix_error (error, "failed to get version number: ");
+		return FALSE;
+	}
+
+	/* set version */
+	version = g_strdup_printf ("%u.%u", buf_rep[1], buf_rep[2]);
+	fu_device_set_version (FU_DEVICE (self), version);
+	return TRUE;
+}
+
+static gboolean
+fu_vli_usbhub_realtek_device_setup (FuDevice *device, GError **error)
+{
+	FuVliUsbhubRealtekDevice *self = FU_VLI_USBHUB_REALTEK_DEVICE (device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* get version */
+	locker = fu_device_locker_new_full (device,
+					    (FuDeviceLockerFunc) fu_device_detach,
+					    (FuDeviceLockerFunc) fu_device_attach,
+					    error);
+	if (locker == NULL)
+		return FALSE;
+	if (!fu_vli_usbhub_realtek_ensure_version_unlocked (self, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_vli_usbhub_device_realtek_detach_raw (FuVliUsbhubRealtekDevice *self, GError **error)
+{
+	FuVliUsbhubDevice *parent = FU_VLI_USBHUB_DEVICE (fu_device_get_parent (FU_DEVICE (self)));
+	guint8 buf[1] = { 0x03 };
+	if (!fu_vli_usbhub_device_i2c_write (parent, 0x6A, 0x31, buf, sizeof(buf), error)) {
+		g_prefix_error (error, "failed to detach: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_vli_usbhub_device_realtek_detach_cb (FuDevice *device,
+					gpointer user_data,
+					GError **error)
+{
+	FuVliUsbhubRealtekDevice *self = FU_VLI_USBHUB_REALTEK_DEVICE (device);
+	guint8 status = 0xff;
+	if (!fu_vli_usbhub_device_realtek_detach_raw (self, error))
+		return FALSE;
+	if (!fu_vli_usbhub_device_realtek_read_status_raw (self, &status, error))
+		return FALSE;
+	if (status != 0x11) {
```

Comment:
this and other status codes I think should be in definitions

Reply:
I've got some more data back from Realtek, so I've added a ton of them to the code.